mteb/abstasks/mteb_result.py

```diff
@@ -0,0 +1,183 @@
+from __future__ import annotations
+
+import json
+from functools import lru_cache
+from pathlib import Path
+from typing import Any, Callable, Dict, Type
+
+import numpy as np
+from pydantic import BaseModel, field_validator
+from TaskMetadata import ISO_LANGUAGE, ISO_LANGUAGE_SCRIPT, LANGUAGES, HFSubset
+
+from mteb.abstasks import AbsTask
+from mteb.abstasks.languages import LangScriptFilter
+from mteb.overview import create_task_list
+
+Split = str
+Scores = Dict[str, Any]
+
+
+@lru_cache(maxsize=None)
+def create_name_to_task_mapping() -> dict[str, Type[AbsTask]]:
+    tasks = create_task_list()
+    return {cls.metadata.name: cls for cls in tasks}
+
+
+def eval_langs_as_dict(
+    eval_langs: LANGUAGES,
+) -> dict[HFSubset, list[ISO_LANGUAGE_SCRIPT]]:
+    if isinstance(eval_langs, list):
+        return {"": eval_langs}
+    return eval_langs  # type: ignore
+
+
+class MTEBResults(BaseModel):
```

Comment:
The main change in this PR. 

Currently, if a dataset has no subset on huggingface a fake subset is added (""). This makes the format standard, but isn't the prettiest I can imagine. An alternative is to use None, which I might prefer, but when serializing it to json None turns to "null". A third option is to convert it to a list of dicts where the dict have 

```
{"train": [{"hf_subset": "en-de" "main_score": 0.5", ...}, {"hf_subset": "en-fr" "main_score": 0.6", ...}, ]
```
This leads to notably less confusion when `"hf_subset": "null"` as the name is specified.

Reply:
Looking this over again I am probably more for the last option.

Reply:
I thought the idea was to include the language even when it's monolingual (no subset case) like in https://github.com/embeddings-benchmark/mteb/issues/639#issuecomment-2097788082.

Reply:
Yep. The issue I found was that we might loose "performance on split hf_subset X" which might be relevant as it is not recoverable from languages. However with the functionality as it is now we can easily change to use languages instead (which might be a better solution)

Reply:
I will wait to get some more opinions on this

Reply:
(@imenelydiaker, @x-tabdeveloping)

Reply:
Isn't it just called `"default"` if you don't specify any subset when uploading the dataset? I think we could totally use that.

Reply:
Apparently, both works (default is probably better):
<img width="340" alt="Screenshot 2024-05-10 at 12 30 43" src="https://github.com/embeddings-benchmark/mteb/assets/23721977/07b4bb24-1438-40c9-a1ab-54534a102c80">

@x-tabdeveloping Would you prefer the hf_subsets or just having language codes instead?

Reply:
We should definitely have `hf_subsets`. I really like the way it is handled in multilingual tasks:
```
{"en": ["eng-Latn"], "et-hu": ["est-Latn", "hun-Latn"]}
```
And then we should follow a similar format with the monolingual tasks just with the default split:
```
{"default": ["eng-Latn"]}
```
In theory only internals should care what the splits are named and its not something we should rely on beyond practicalities.

Reply:
I certainly like the first option with the list. I believe adding the languages is a good idea, as we can't (or shouldn't) rely on the splits being language names.

Reply:
Now uses a list